scylla/tests/ccm-integration/logged_cli.rs

```diff
@@ -0,0 +1,344 @@
+use anyhow::{Context, Error};
+use std::collections::HashMap;
+use std::ops::Deref;
+use std::process::{ExitStatus, Stdio};
+use std::sync::atomic::AtomicI32;
+use std::sync::Arc;
+use tokio::fs::{File, OpenOptions};
+use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
+use tokio::process::Command;
+use tokio::sync::{Mutex};
+use tokio::task;
+
+pub(crate) struct LoggedCmd {
+    log_file: String,
+    file: Option<Arc<Mutex<File>>>,
+    run_id: AtomicI32,
+}
+
+#[macro_export]
+macro_rules! run_options {
+    ($($key:ident = $value:expr),* $(,)?) => {
+        Some(RunOptions {
+            $($key: $value,)*
+            ..Default::default()
+        })
+    };
+}
+
+#[derive(Default, Debug)]
+pub(crate) struct RunOptions {
+    pub env: HashMap<String, String>,
+    pub allow_failure: Option<bool>,
+}
+
+impl LoggedCmd {
+    pub(crate) fn new() -> Self {
+        LoggedCmd {
+            log_file: "".to_string(),
+            file: None,
+            run_id: AtomicI32::new(1),
+        }
+    }
+
+    pub(crate) async fn set_log_file(&mut self, file_name: String) -> Result<(), Error> {
+        self.log_file = file_name.clone();
+        let file = OpenOptions::new()
+            .create(true)
+            .append(true)
+            .open(self.log_file.as_str())
+            .await
+            .context(format!("failed to open ccm log file {}", self.log_file))?;
+        self.file = Some(Arc::new(Mutex::new(file)));
+        Ok(())
+    }
+
+    pub(crate) async fn run_command(
+        &self,
+        command: &str,
+        args: &[&str],
+        opts: Option<RunOptions>,
+    ) -> Result<ExitStatus, Error> {
```

Comment:
:wrench: This function is too long. Please extract helper functions from it, so that it's easier to follow its logic.